File: vtu-exam-prep/scripts/process_uhv_mcqs.mjs
import fs from 'fs';
import path from 'path';
import { createClient } from '@supabase/supabase-js';
import dotenv from 'dotenv';

dotenv.config({ path: '.env.local' });
dotenv.config({ path: '.env' });

const supabaseUrl = process.env.VITE_SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SECRET_KEY || process.env.VITE_SUPABASE_PUBLISHABLE_KEY;

if (!supabaseUrl || !supabaseKey) {
  console.error('Missing Supabase credentials');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);

function makeExp(optKey, ans, optText, correctText) {
  if (optKey === ans) return "Correct. '" + optText + "' is the right answer.";
  return "Incorrect. The correct answer is " + ans + ") " + correctText + ".";
}

function parseMarkdownFile(filepath, sourceName) {
  console.log('    Reading file...');
  const content = fs.readFileSync(filepath, 'utf-8');
  console.log('    File read OK. Length: ' + content.length);

  // Detect format: does it have "> **Answer:" lines?
  const hasAnswerLines = content.includes('> **Answer:');
  console.log('    Format: ' + (hasAnswerLines ? 'Answer block' : 'Inline checkmark'));

  const questions = [];

  if (hasAnswerLines) {
    // Format 1: > **Answer: X**
    const pattern = /\*\*(\d+)\.\*\*\s*([\s\S]*?)\n\n-\s*A\)\s*(.*)\n-\s*B\)\s*(.*)\n-\s*C\)\s*(.*)\n-\s*D\)\s*(.*)\n\n>\s*\*\*Answer:\s*([A-D])\*\*/g;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const [, , qText, optA, optB, optC, optD, ans] = match;
      const opts = { A: optA.trim(), B: optB.trim(), C: optC.trim(), D: optD.trim() };
      questions.push({
        source_type: 'notes', source_name: sourceName,
        question_text: qText.trim(),
        option_a: opts.A, option_b: opts.B, option_c: opts.C, option_d: opts.D,
        correct_answer: ans.trim(),
        explanation_a: makeExp('A', ans, opts.A, opts[ans]),
        explanation_b: makeExp('B', ans, opts.B, opts[ans]),
        explanation_c: makeExp('C', ans, opts.C, opts[ans]),
        explanation_d: makeExp('D', ans, opts.D, opts[ans]),
        frequency: 1,
      });
    }
  } else {
    // Format 2: inline checkmark
    // Split by question headers
    const qPattern = /\*\*(\d+)\.\*\*/g;
    const positions = [];
    let m;
    while ((m = qPattern.exec(content)) !== null) {
      positions.push({ index: m.index, num: m[1] });
    }

    for (let i = 0; i < positions.length; i++) {
      const start = positions[i].index;
      const end = i + 1 < positions.length ? positions[i + 1].index : content.length;
      const block = content.substring(start, end);

      // Get question text (between **N.** and first option)
      const qTextMatch = block.match(/\*\*\d+\.\*\*\s*([\s\S]*?)(?=\n\s*-\s*[A-D]\))/);
      if (!qTextMatch) continue;

      const qText = qTextMatch[1].trim();
      const opts = {};
      let correctAnswer = null;

      // Extract each option line
      const optLines = block.match(/-\s*[A-D]\)\s*.+/g);
      if (!optLines || optLines.length < 4) continue;

      for (const line of optLines) {
        const optMatch = line.match(/-\s*([A-D])\)\s*(.*)/);
        if (!optMatch) continue;
        const key = optMatch[1];
        let text = optMatch[2].trim();
        // Remove checkmark and mark as correct
        if (text.includes('\u2705') || text.includes('\u2713')) {
          correctAnswer = key;
          text = text.replace(/[\u2705\u2713]/g, '').trim();
        }
        opts[key] = text;
      }

      if (!opts.A || !opts.B || !opts.C || !opts.D || !correctAnswer) continue;

      questions.push({
        source_type: 'notes', source_name: sourceName,
        question_text: qText,
        option_a: opts.A, option_b: opts.B, option_c: opts.C, option_d: opts.D,
        correct_answer: correctAnswer,
        explanation_a: makeExp('A', correctAnswer, opts.A, opts[correctAnswer]),
        explanation_b: makeExp('B', correctAnswer, opts.B, opts[correctAnswer]),
        explanation_c: makeExp('C', correctAnswer, opts.C, opts[correctAnswer]),
        explanation_d: makeExp('D', correctAnswer, opts.D, opts[correctAnswer]),
        frequency: 1,
      });
    }
  }

  return questions;
}

async function main() {
  console.log('============================================================');
  console.log('UHV MCQ Processing Script');
  console.log('============================================================');

  console.log('\n[STEP 1] Processing Notes MCQs...');
  const notesDir = 'C:/4thsem/4thsem_finals/uhv/notes_mcq';
  const mdFiles = fs.readdirSync(notesDir).filter(f => f.endsWith('.md')).sort();

  let allQuestions = [];
  for (const filename of mdFiles) {
    const filepath = path.join(notesDir, filename);
    const sourceName = filename.replace('.md', '');
    console.log('  Processing ' + sourceName + '...');
    try {
      const questions = parseMarkdownFile(filepath, sourceName);
      console.log('    Parsed ' + questions.length + ' questions');
      allQuestions = allQuestions.concat(questions);
    } catch (err) {
      console.error('    ERROR parsing ' + sourceName + ': ' + err.message);
    }
  }

  console.log('\n  Total notes questions: ' + allQuestions.length);

  if (allQuestions.length === 0) {
    console.log('  No questions to insert. Exiting.');
    return;
  }

  console.log('  Inserting into Supabase (batch)...');

  for (let i = 0; i < allQuestions.length; i += 50) {
    const batch = allQuestions.slice(i, i + 50);
    try {
      const { error } = await supabase.from('uhv_questions').insert(batch);
      if (error) {
        console.error('  Batch ' + (Math.floor(i / 50) + 1) + ' FAILED: ' + error.message);
      } else {
        console.log('  Inserted batch ' + (Math.floor(i / 50) + 1) + ' (' + batch.length + ' rows)');
      }
    } catch (err) {
      console.error('  Network error on batch ' + (Math.floor(i / 50) + 1) + ': ' + err.message);
    }
  }

  console.log('\n[OK] Notes MCQs done.');

  // PDFs skipped - need Vision API
  console.log('\n[STEP 2] PDF Question Papers');
  console.log('  Skipped - scanned PDFs need a valid OpenAI API key for Vision extraction.');

  console.log('\n============================================================');
  console.log('Done! Check your Supabase uhv_questions table.');
}

main().catch(err => {
  console.error('Fatal:', err);
  process.exit(1);
});

File: vtu-exam-prep/scripts/process_uhv_mcqs.py
import os
import re
import json
import fitz  # PyMuPDF
import base64
from supabase import create_client, Client
from dotenv import load_dotenv

# Load env variables
load_dotenv('.env.local')
load_dotenv('.env')

supabase_url = os.environ.get("VITE_SUPABASE_URL")
supabase_key = os.environ.get("SUPABASE_SECRET_KEY")

if not supabase_url or not supabase_key:
    print("Missing Supabase credentials. Check .env")
    exit(1)

supabase: Client = create_client(supabase_url, supabase_key)

def generate_explanation(question_text, option_text, option_key, correct_answer, correct_text):
    """Generate a simple but helpful explanation for each option."""
    if option_key == correct_answer:
        return f"Correct. '{option_text}' is the right answer to this question."
    else:
        return f"Incorrect. The correct answer is {correct_answer}) {correct_text}."

def process_markdown_file(filepath, source_name):
    print(f"\nProcessing {filepath}...")
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    pattern = re.compile(
        r'\*\*(\d+)\.\*\*\s*(.*?)\n\n'
        r'-\s*A\)\s*(.*?)\n'
        r'-\s*B\)\s*(.*?)\n'
        r'-\s*C\)\s*(.*?)\n'
        r'-\s*D\)\s*(.*?)\n\n'
        r'>\s*\*\*Answer:\s*([A-D])\*\*',
        re.DOTALL
    )
    matches = pattern.findall(content)
    
    print(f"Found {len(matches)} questions in {source_name}.")
    
    inserted = 0
    for match in matches:
        q_num, q_text, opt_a, opt_b, opt_c, opt_d, ans = match
        q_text = q_text.strip()
        opt_a = opt_a.strip()
        opt_b = opt_b.strip()
        opt_c = opt_c.strip()
        opt_d = opt_d.strip()
        ans = ans.strip()
        
        options = {'A': opt_a, 'B': opt_b, 'C': opt_c, 'D': opt_d}
        correct_text = options[ans]
        
        row = {
            "source_type": "notes",
            "source_name": source_name,
            "question_text": q_text,
            "option_a": opt_a,
            "option_b": opt_b,
            "option_c": opt_c,
            "option_d": opt_d,
            "correct_answer": ans,
            "explanation_a": generate_explanation(q_text, opt_a, 'A', ans, correct_text),
            "explanation_b": generate_explanation(q_text, opt_b, 'B', ans, correct_text),
            "explanation_c": generate_explanation(q_text, opt_c, 'C', ans, correct_text),
            "explanation_d": generate_explanation(q_text, opt_d, 'D', ans, correct_text),
            "frequency": 1
        }
        
        try:
            supabase.table('uhv_questions').insert(row).execute()
            inserted += 1
        except Exception as e:
            print(f"  Failed Q{q_num}: {e}")
    
    print(f"  [OK] Inserted {inserted}/{len(matches)} questions from {source_name}.")


def process_pdf_file(filepath, source_name):
    print(f"\nProcessing PDF {filepath}...")
    doc = fitz.open(filepath)
    
    full_text = ""
    for page_num in range(len(doc)):
        page = doc.load_page(page_num)
        text = page.get_text("text")
        full_text += text + "\n"
    
    if len(full_text.strip()) < 100:
        print(f"  [WARN] PDF {source_name} has little/no extractable text (scanned image?). Skipping.")
        return
    
    # Try multiple patterns to extract MCQs from the PDF text
    # VTU papers typically have formats like:
    # 1. Question text
    # a) option  b) option  c) option  d) option
    # OR
    # 1. Question text
    # A. option  B. option  C. option  D. option
    
    # Clean up the text
    full_text = full_text.replace('\r\n', '\n').replace('\r', '\n')
    
    # Pattern 1: numbered question with a/b/c/d options
    pattern1 = re.compile(
        r'(\d+)\.\s*(.*?)\s*'
        r'[aA][\.\)]\s*(.*?)\s*'
        r'[bB][\.\)]\s*(.*?)\s*'
        r'[cC][\.\)]\s*(.*?)\s*'
        r'[dD][\.\)]\s*(.*?)(?=\n\d+\.|$)',
        re.DOTALL
    )
    
    matches = pattern1.findall(full_text)
    
    if not matches:
        print(f"  [WARN] Could not parse MCQs from {source_name} text. Saving raw text for manual review.")
        with open(f"scripts/pdf_text_{source_name}.txt", 'w', encoding='utf-8') as f:
            f.write(full_text)
        print(f"  Saved raw text to scripts/pdf_text_{source_name}.txt")
        return
    
    print(f"  Found {len(matches)} questions in {source_name}.")
    
    inserted = 0
    for match in matches:
        q_num, q_text, opt_a, opt_b, opt_c, opt_d = match
        q_text = ' '.join(q_text.split()).strip()
        opt_a = ' '.join(opt_a.split()).strip()
        opt_b = ' '.join(opt_b.split()).strip()
        opt_c = ' '.join(opt_c.split()).strip()
        opt_d = ' '.join(opt_d.split()).strip()
        
        if not q_text or not opt_a or not opt_b:
            continue
        
        # Default to 'A' since we can't determine the answer from the PDF
        # This will need manual correction or a future AI pass
        ans = 'A'
        
        row = {
            "source_type": "qp",
            "source_name": source_name,
            "question_text": q_text,
            "option_a": opt_a,
            "option_b": opt_b,
            "option_c": opt_c,
            "option_d": opt_d,
            "correct_answer": ans,
            "explanation_a": "Answer extracted from previous year paper. Verify with your notes.",
            "explanation_b": "Answer extracted from previous year paper. Verify with your notes.",
            "explanation_c": "Answer extracted from previous year paper. Verify with your notes.",
            "explanation_d": "Answer extracted from previous year paper. Verify with your notes.",
            "frequency": 1
        }
        
        try:
            supabase.table('uhv_questions').insert(row).execute()
            inserted += 1
        except Exception as e:
            print(f"  Failed Q{q_num}: {e}")
    
    print(f"  [OK] Inserted {inserted} questions from {source_name}.")


# ─── MAIN ──────────────────────────────────────────────
print("=" * 60)
print("UHV MCQ Processing Script (No OpenAI Required)")
print("=" * 60)

# Step 1: Process Notes MCQs (Markdown files)
print("\n[STEP 1] Processing Notes MCQs...")
notes_dir = 'C:/4thsem/4thsem_finals/uhv/notes_mcq'
for filename in sorted(os.listdir(notes_dir)):
    if filename.endswith('.md'):
        process_markdown_file(os.path.join(notes_dir, filename), filename.replace('.md', ''))

print("\n[OK] All Notes MCQs processed.")

# Step 2: Process QP PDFs
print("\n[STEP 2] Processing Question Paper PDFs...")
qp_dir = 'C:/4thsem/4thsem_finals/uhv/qp_uhv'
for filename in sorted(os.listdir(qp_dir)):
    if filename.endswith('.pdf'):
        process_pdf_file(os.path.join(qp_dir, filename), filename.replace('.pdf', ''))

print("\n[OK] All QP PDFs processed.")
print("=" * 60)
print("Done! Check your Supabase uhv_questions table.")
